### ISSUE TITLE
UPNP not obeying "bind-address-ipv4" setting

### ISSUE THREAD

**Issue by robross0606** [transmission/transmission#804]
Even with bind-address-ipv4 setting set properly, UPNP packets are using the wrong interface.

**Comment by robross0606** [comment]
Even with bind-address-ipv4 configured properly (and used for other packets), UPNP packets are "leaking" onto other interfaces.

**Comment by alexander-gabriel** [comment]
Some wrote and submitted a patch for this 3 years ago.
https://trac.transmissionbt.com/ticket/5990

**Comment by robross0606** [comment]
This is precisely the kind of issue (and patch) that shows Transmission development leadership is missing.  People are trying to contribute, but nothing is being done to either resolve known issues or encourage active development .

**Comment by LaserEyess** [comment]
@robross0606 that patch, by the author's own admission, is incomplete. It doesn't handle ipv6 or even fix nat-pmp. Plus he submitted it on the deprecated trac. It's not hard to realize why no one ever paid attention to it, they didn't even know it existed. 

I rebased the patch and made a pull request #845, feel free to test with it. But it shouldn't be merged until IPv6 works, and I have to figure that out.

**Comment by LaserEyess** [comment]
@robross0606 are you willing to test if #845 solves your issue? If you can't build it on your own I can provide you a build.

**Comment by robross0606** [comment]
Absolutely, but it sounds like there are some caveats as to what should be tested. What exactly is supposed to be working with this branch?

**Comment by LaserEyess** [comment]
It will only work for ipv4 upnp, not nat-pmp. And it does not work for ipv6. The logic being that ipv6 does not really need NAT since you should have a local /64 to use. I will look into nat-pmp, I just want to see if this will work with upnp right now.

**Comment by robross0606** [comment]
I have IPv6 disabled anyway.  I'll try to set up and give it a test at some point, but probably not until next weekend.

**Comment by robross0606** [comment]
> @robross0606 are you willing to test if #845 solves your issue? If you can't build it on your own I can provide you a build.

I tried but was unable to get this to build without blowing up my system with a bunch of dev libraries and custom configuration.  I don't have a dedicated dev system where I can set up the necessary toolchain.  If you can provide a drop-in build, I will test it.  Sorry.

**Comment by robross0606** [comment]
I was able to get ```cmake ..``` to work, but the subsequent ```make``` call fails:
```
Scanning dependencies of target event
[  1%] Creating directories for 'event'
[  1%] Performing download step (DIR copy) for 'event'
[  1%] No patch step for 'event'
[  2%] No update step for 'event'
[  3%] Performing configure step for 'event'
/bin/sh: 1: /home/torrent/Downloads/transmission-upnp_bind_address/build/third-party/event/src/event/autogen.sh: not found
make[2]: *** [CMakeFiles/event.dir/build.make:108: third-party/event/src/event-stamp/event-configure] Error 127
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/event.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```
It is unclear what went wrong.

**Comment by robross0606** [comment]
Is there a blessed list of "dev" libraries needed in Ubuntu to get this built?  That would surely make life much easier.

**Comment by robross0606** [comment]
I've managed to whittle the build issues down to:
```
-- Could NOT find DHT (missing: DHT_LIBRARY DHT_INCLUDE_DIR)
-- Could NOT find UTP (missing: UTP_LIBRARY UTP_INCLUDE_DIR)
-- Could NOT find B64 (missing: B64_LIBRARY B64_INCLUDE_DIR)
```
And the subsequent ```make``` error:
```
CMake Error: The source directory "/home/torrent/Downloads/transmission-upnp_bind_address/build/third-party/dht/src/dht" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
make[2]: *** [CMakeFiles/dht.dir/build.make:108: third-party/dht/src/dht-stamp/dht-configure] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/dht.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```
The list of required dependencies for Ubuntu build on the wiki appears to be incorrect.  I have not found anything about ```dht``` mentioned on the wiki.

**Comment by robross0606** [comment]
Good grief.  After searching everywhere, I found that the Git repo has submodules to pull down the necessary code.  If that's documented someplace in the wiki, I totally missed it.  :-/

**Comment by robross0606** [comment]
Okay, so transmission-cli is not part of the package.  Another lesson learned.  ```transmission-daemon``` appears to return the expected version:
```
transmission-daemon 2.94+ (c16701bb5b)
```

**Comment by robross0606** [comment]
I don't know exactly what this *did* but it doesn't appear to be working.  The UPNP packets do not appear to be going anywhere now.  Not across my regular network where they were "leaking" before and not across the bound interface either.

**Comment by robross0606** [comment]
Correction, they still appear to have leaked onto the regular network.  I forgot that I had a deny filter on my UPNP server to prevent the "leak" from working.  When I removed that filter, the port opened on my regular network again.  UPNP still appears to be ignoring the

**Comment by robross0606** [comment]
I think the problem is NAT-PMP.  There doesn't appear to be a way to disable NAT-PMP in Transmission.  Without NAT-PMP also going across the bound interface, the torrent ports remain "closed".

**Comment by robross0606** [comment]
This is definitely the problem.  
```
[2019-03-23 12:39:59.220] Port Forwarding (NAT-PMP) Found public address "###.###.###.###" (/home/torrent/Code/transmission/libtransmission/natpmp.c:138)
```
Transmission is still finding my public address via NAT-PMP even though the IPV4 address is bound to an interface with an entirely different public address:

**Comment by robross0606** [comment]
This version also seems to have a strange problem where transmission-daemon isn't stopping and starting in a timely manner.  I have had to manually kill the executable a couple times.  Never had that problem before.  Very strange.

**Comment by robross0606** [comment]
This version also appears to have broken the ability to NOT bind to an address.  If I leave the bind-address-ipv4 empty, port forwarding hangs and eventually crashes out the daemon.  Reverting to regular version.

**Comment by mikedld** [comment]
@robross0606, this is not IM or Twitter or whatever, please keep the number of comments low and their quality high as much as you can --- each of them results in an email being sent to a number of people. Real-time updates aren't necessary.

**Comment by robross0606** [comment]
If you look, the number of resulting comments is low.  The number of edits was high.  If the build process worked properly and the wiki was updated, there would be no confusion and very low comments.  I edited the results as they happened.  Flaming people who are spending the entire afternoon helping to test the software is *not* a great way to facilitate a thriving community, which is exactly the problem Transmission is having.  And, might I add that your email was not even remotely "high quality" since it added a non-productive and toxic tone to an otherwise completely productive thread.

**Comment by LaserEyess** [comment]
> This version also appears to have broken the ability to NOT bind to an address. If I leave the `bind-address-ipv4` empty, port forwarding hangs and eventually crashes out the daemon. Reverting to regular version.

That's certainly sounds like a bug, I can add a check to make sure bind-address-ipv4 exists. Or maybe it just doesn't like the default, 0.0.0.0.

> Correction, they still appear to have leaked onto the regular network. I forgot that I had a deny filter on my UPNP server to prevent the "leak" from working. When I removed that filter, the port opened on my regular network again. UPNP still appears to be ignoring the bind setting.

> I think the problem is NAT-PMP. There doesn't appear to be a way to disable NAT-PMP in Transmission. Without NAT-PMP also going across the bound interface, the torrent ports remain "closed".

The issue with NAT-PMP as I read it is that it requires you give a default gateway. I can probably figure out how to make that work in linux,  but certainly not macOS/Windows (at least on my own). It would require a lot more complexity so I decided to leave it as is. If that's what's causing your issue I don't think I can fix it. However, focusing on the UPNP for right now: are you absolutely sure it's still leaking, even with bind-address-ipv4 set? Or can you just not tell vs NAT-PMP?

**Comment by LaserEyess** [comment]
@robross0606 I took another look at this, and from my testing it is working as expected. Specifically: 
- There is no leaking on interfaces that aren't assigned the ip in `bind-address-ipv4`
- I get no crashes or any abnormal behavior when `bind-address-ipv4` is set to any value, even an invalid ipv4 address (it falls back to INADDR_ANY)

Can you give some more detailed steps on how your setup crashed?
